### PR TITLE
feat: react library

### PR DESCRIPTION
**Pull request by lharries**
Adds a react library which can be used by create react apps and next js

API heavily inspired by posthog's react native app

Introduces the provider:
`<PostHogProvider>...</PostHogProvider>` which can either take an initiated posthog client, or an API key and options

And the following hooks:
- usePostHog
- useFeatureFlag
- useFeatureFlagEnabled
- useFeatureFlagPayload
- useFeatureFlags

TODO:
- [x] Update the playground to use these libraries
- [x] Add the feature flag playload hook
- [x] Bundle it properly with the main package e.g. `import { useFeatureFlag } from 'posthog-js/react' @benjackwhite 
- [x] Get team experiments to review the feature flags API

After merging:
- [ ] Update the docs

## Checklist
- [x] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers

### CODE REVIEW

**Comment by lharries** [comment]
@benjackwhite would be great to get your input on the bundling. Feel free to ignore the files outside of the '/react' folder.

**Comment by lharries** [comment]
@ivanagas here's the plan for the new react API:
- using feature flags: https://github.com/PostHog/posthog-js/pull/540/files#diff-3f9639244c4f35263ae8dca85e6e0192116557ced4ea7f4afadb2ca764a912fd
- doing events: https://github.com/PostHog/posthog-js/pull/540/files#diff-c150c8de1585c706c7d7ec0d0d566ba6a2ab92394c5782123820f190f9869b4f

**Comment by lharries** [comment]
The react native API has the hooks `getFeatureFlag` which is the same as `isFeatureEnabled` in the js library. And `getFeatureFlags` which is the same as the `onFeatureFlags` callback

However, with the payload function too I wonder if `getFeatureFlag` will be confusing for the JS SDK. In this PR I've proposed:
- useFeatureFlagEnabled
- useFeatureFlagPayload
- useFeatureFlags

@neilkakkar @EDsCODE and @liyiy what API do you want for the react feature flags?

**Comment by benjackwhite** [comment]
Made some changes here to simplify some things and build the output so that one can import from `posthog-js/react`.

One contentious thing: Should we offer `getFeatureFlags` ? It feels like a possible anti-pattern over `getFeatureFlag` which is more manageable.

File: react/src/hooks/useFeatureFlag.ts
Comment (benjackwhite) [inline react/src/hooks/useFeatureFlag.ts:26]: Hot take - I think we should _only_ support this code path.

`useFeatureFlagEnabled` is the same as writing
```
const isEnabled = !!useFeatureFlag("flag")
```

`useFeatureFlags` avoids the callback code that we use for Experiments so is likely to get someone into a state they didn't intend.

`useFeatureFlagPayload` is identical to `useFeatureFlag` but with the case that it will respond with undefined if the key isn't a payload... I think this is super confusing. If I was the dev implementing this I wouldn't understand the difference between those two functions and would likely implement the wrong one.

Comment (neilkakkar) [reply react/src/hooks/useFeatureFlag.ts:26]: I don't like this, let's chat sync, sent an invite for Wed!

Comment (mariusandra) [reply react/src/hooks/useFeatureFlag.ts:26]: I don't have a strong stance on here (and we can iterate), though there's another thing.

I didn't run the code, but looking at [Github](https://github.com/PostHog/posthog-js/blob/master/src/posthog-featureflags.ts#L276), `client.onFeatureFlags` doesn't return anything. So I'm not sure what's this `return client.onFeatureFlags()` returning.

From useEffect we should return a function that removes the feature flag listener. Otherwise when this component unmounts, posthog will still be calling the callback, and through it a useState setter on an unmounted component. Unfortunately posthog-js doesn't seem to support both attaching and detaching feature flag callbacks 🤔

Comment (benjackwhite) [reply react/src/hooks/useFeatureFlag.ts:26]: Ah this was probably a copy-paste from `posthog-js-lite`

**Comment by benjackwhite** [comment]
Thought about this a little more (posting as pre-amble for our meeting about it):

If we were to implement this as a component what would it look like? Thinking it through reinforced my idea that a single method approach makes sense.

```tsx
// GENERAL
// Match if value is truthy
<PostHogFeature flag="my-boolean-flag">
  <p>New feature!!</p>
</PosthogFeature>


// BOOL
// Match if value is explicitly what is set (boolean)
<PostHogFeature flag="my-boolean-flag" match={true}>
  <p>New feature!!</p>
</PosthogFeature>
// OR get the raw value to render
<PostHogFeature flag="my-boolean-flag">
  {payload => payload ? <p>Feature enabled</p> : <p>Not enabled</p>}
</PosthogFeature>


// STRING
// Match if value is explicitly what is set (string)
<PostHogFeature flag="my-string-flag" match={"test"}>
  <p>Test feature!!</p>
</PosthogFeature>
// OR get the raw value to render
<PostHogFeature flag="my-string-flag">
  {payload => payload === "test" 
    ? <p>Feature enabled</p> 
    : payload === "control"  
    ? <p>Not enabled</p>
    : null
    }
</PosthogFeature>

// JSON
// For this case one would almost always want the payload to render with an optional variantId
<PostHogFeature flag="my-payload-flag">
    {(payload, variantId) => (
        <p>New feature: {payload.name}!!</p>
        <a href={payload.cta}>CTA</a>
    )}
</PosthogFeature>
```

**Comment by lharries** [comment]
@neilkakkar I think we are currently blocked by your take on what the react api should be for feature flags. Would be great to get this merged in as makes the experience much better for react and nextjs users

**Comment by neilkakkar** [comment]
To unblock you for now, for this specific library, what Ben proposed above makes sense to me (as long as we're not changing the posthog-js api): https://github.com/PostHog/posthog-js/pull/540#issuecomment-1438216439

with a few modifications:

1. Get rid of the BOOL case.
2. For the STRING case: The `match={}` should match the variant key instead of the payload in the string case.
3. JSON case renames `variantId` to `variantKey`. And this is the common way to get 'raw value to render' for all the above.


If we do change this later on, would be happy to update this library myself :) 

---

And for the hooks:

useFeatureFlag - returns the variant key, which uses `posthog.getFeatureFlag()`. (*this is different from what you suggested*)

useFeatureFlagEnabled - which uses isFeatureEnabled
useFeatureFlagPayload - which uses getFeatureFlagPayload

Not sure if `useFeatureFlags` will be useful here, that seems confusing, what was the intention behind it?

**Comment by lharries** [comment]
@neilkakkar thanks! 

`useFeatureFlag` - I'm going to suggest we leave this out for now. The multivariate case is quite confusing! It would return the payload if it exists but otherwise a boolean

`useFeatureFlags` - in quite a few things I've built I've found it useful to iterate over the feature flags so going to suggest we leave it in for now

**Comment by neilkakkar** [comment]
Mind explaining re: `useFeatureFlag`, why is returning the variant key more confusing that returning the payload or the variant sometimes? 🤔

**Comment by neilkakkar** [comment]
re: useFeatureFlags, it sounds like it's goal is to return all flags, so I'd suggest following the server-side lib convention for returning all flags, and calling it `useAllFlags` ? This also makes the names different enough to suggest their use case better :)

File: react/src/hooks/useFeatureFlag.ts
Comment (lharries) [reply react/src/hooks/useFeatureFlag.ts]: > From useEffect we should return a function that removes the feature flag listener. Otherwise when this component unmounts, posthog will still be calling the callback, and through it a useState setter on an unmounted component. Unfortunately posthog-js doesn't seem to support both attaching and detaching feature flag callbacks 🤔

I've added removal of the onFeatureFlags listener when the component is unmounted

**Comment by lharries** [comment]
> re: useFeatureFlags, it sounds like it's goal is to return all flags, so I'd suggest following the server-side lib convention for returning all flags, and calling it useAllFlags ? This also makes the names different enough to suggest their use case better :)

Doesn't it return just the active flags rather than all flags? (I think we had the problem with the regression in the user interview app related to this)

**Comment by neilkakkar** [comment]
Sounds good, yes even better 😅

**Comment by lharries** [comment]
> Mind explaining re: useFeatureFlag, why is returning the variant key more confusing that returning the payload or the variant sometimes? 🤔

I think it links to the issues that Ben and I were mentioning

For the `useFeatureFlag`, if I was calling it and had a payload I'd expect it to return a payload - regardless of whether it's a multivariate or a boolean.

**Comment by neilkakkar** [comment]
Hmm, removing it means we end up missing a big piece of usefulness here, since right now there is no equivalent of `posthog.getFeatureFlag()`. You can somewhat get it from `getActiveFeatureFlags`, but that's missing the `$feature_flag_called` calls for experiments & audit.

I'll leave it up to you though!

**Comment by lharries** [comment]
> Hmm, removing it means we end up missing a big piece of usefulness here, since right now there is no equivalent of posthog.getFeatureFlag(). You can somewhat get it from getActiveFeatureFlags, but that's missing the $feature_flag_called calls for experiments & audit.

Good point. I like the idea of a `useFeatureFlagKey` function or a `useFeatureFlagVariantId` however `getFeatureFlag` seems to return either a boolean or a string. I would have expected the key to always be a string?

**Comment by neilkakkar** [comment]
Yeah, `useFeatureFlagVariantKey` is more accurate, instead of FeatureFlagKey which means something else.

The boolean is the variant key for a flag that has no variants (and responds with true), which exists for backwards compatibility iirc, but yeah for a flag with no variants you really shouldn't be calling `getFeatureflag()`

**Comment by lharries** [comment]
> The boolean is the variant key for a flag that has no variants (and responds with true), which exists for backwards compatibility iirc, but yeah for a flag with no variants you really shouldn't be calling getFeatureflag()

I've just tried it and for a boolean value getFeatureFlag returns `true` if it's a boolean type. I'd have thought that the variantKey would be a string? I'm skeptical we'd want to return true/false for the VariantKey and have boolean in the type signature

I'd be keen to merge what we have so far and then in parallel I think you're still working through the proposal ben mentioned to streamline it to just getFeatureFlagPayload and getFeatureFlagVariantKey (unless you have reached a decision)

**Comment by lharries** [comment]
@neilkakkar if you agree, would you mind giving a final review of the PR including a check of the new removing event listener function I added

**Review by neilkakkar — APPROVED** [review]

**Comment by neilkakkar** [comment]
Didn't manually try it out, but I'm hoping you have! Code looks good

**Comment by lharries** [comment]
@neilkakkar based on our discussion I agree and it makes sense to add useFeatureFlagVariantKey - I've added that here: https://github.com/PostHog/posthog-js/pull/540/commits/7376f328b9c297731dfa0ff3c7de419c2102c935 can you confirm you're happy with it?

**Review by neilkakkar — APPROVED** [review]

**Comment by lharries** [comment]
@ivanagas and @pjhul we now have a react library! Would be awesome to get your help with updating the docs and older tutorials. I'll take a crack at some of them later too

**Comment by lharries** [comment]
@joethreepwood could be good to have some marketing support around this once we've updated the docs. Main benefit for users it that it makes implementation much easier/quicker.

**Comment by joethreepwood** [comment]
Think it can go in the Changelog this week, or is this going to drift to next?

**Comment by lharries** [comment]
I'll update the docs to a good enough place by EOD that it can go in this week

**Comment by joethreepwood** [comment]
> @joethreepwood as discussed @pauldambra has a few more tweaks he wants to do so likely this will be next week instead

I thought that was in reference to Dashboard templates, or is that the case here too?

**Comment by defunctzombie** [comment]
I know its been a while since this was implemented - but I wanted to drop a note that I was recently tripped up by the decision to not set the initial state of the flag:

https://github.com/PostHog/posthog-js/blob/main/react/src/hooks/useFeatureFlagEnabled.ts#L7-L9

This results in users getting undefined and then getting the flag value - even when the flag value is known at startup. I stumbled upon this because of some undesired thrashing and rendering in our app.

IMO the workarounds should be on nextjs users or SSR users to maybe have a separate set of functions - and for everyone that does not care about SSR to not be hit with this behavior.

**Pull request merged**
